source/slang/slang-ir-lower-enum-type.cpp

```diff
@@ -0,0 +1,139 @@
+    IRStructKey* valueKey = nullptr;
+    IRStructKey* hasValueKey = nullptr;
+
+    EnumTypeLoweringContext(IRModule* inModule)
+        : module(inModule), workList(inModule), workListSet(inModule)
+    {
+    }
+
+    struct LoweredEnumTypeInfo : public RefObject
+    {
+        IRType* enumType = nullptr;
+        IRType* loweredType = nullptr;
+    };
+    Dictionary<IRInst*, RefPtr<LoweredEnumTypeInfo>> loweredEnumTypes;
+
+    void addToWorkList(IRInst* inst)
+    {
+        if (workListSet.contains(inst))
+            return;
+
+        workList.add(inst);
+        workListSet.add(inst);
+    }
+
+    LoweredEnumTypeInfo* getLoweredEnumType(IRInst* type)
+    {
+        if (auto loweredInfo = loweredEnumTypes.tryGetValue(type))
+            return loweredInfo->Ptr();
+
+        if (!type)
+            return nullptr;
+
+        if (type->getOp() != kIROp_EnumType)
+            return nullptr;
+
+        RefPtr<LoweredEnumTypeInfo> info = new LoweredEnumTypeInfo();
+        auto enumType = cast<IREnumType>(type);
+        auto valueType = enumType->getTagType();
+        info->enumType = (IRType*)type;
+        info->loweredType = valueType;
+        loweredEnumTypes[type] = info;
+        return info.Ptr();
+    }
+
+    void processEnumType(IREnumType* inst)
+    {
+        auto loweredEnumTypeInfo = getLoweredEnumType(inst);
+        SLANG_ASSERT(loweredEnumTypeInfo);
+        SLANG_UNUSED(loweredEnumTypeInfo);
+    }
+
+    void processEnumCast(IRInst* inst)
+    {
+        IRBuilder builderStorage(module);
+        auto builder = &builderStorage;
+        builder->setInsertBefore(inst);
+
+        auto type = inst->getFullType();
+        auto value = inst->getOperand(0);
+
+        IRType* dstType = type;
+        if (auto enumType = getLoweredEnumType(dstType))
+        { // Cast was into enum, so use tag type instead.
+            dstType = enumType->loweredType;
+        }
+
+        auto cast = builder->emitCast(dstType, value);
+
+        inst->replaceUsesWith(cast);
+        inst->removeAndDeallocate();
+    }
+
+    void processInst(IRInst* inst)
+    {
+        switch (inst->getOp())
+        {
+        case kIROp_EnumType:
+            processEnumType((IREnumType*)inst);
+            break;
+        case kIROp_CastEnumToInt:
+        case kIROp_CastIntToEnum:
+        case kIROp_EnumCast:
+            processEnumCast(inst);
+            break;
+        default:
+            break;
+        }
+    }
+
+    void processModule()
+    {
+        addToWorkList(module->getModuleInst());
+
+        while (workList.getCount() != 0)
+        {
+            IRInst* inst = workList.getLast();
+            workList.removeLast();
+            workListSet.remove(inst);
+
+            processInst(inst);
+
+            for (auto child = inst->getLastChild(); child; child = child->getPrevInst())
+            {
+                addToWorkList(child);
+            }
+        }
+
+        // Replace all optional types with lowered struct types.
```

Comment:
The comments seemed to be copied from another pass and need to be updated here.

Reply:
Whoops, good catch. I had used `slang-ir-lower-optional-type.cpp` as reference.